File: ios/Classes/AlhPdfViewConfiguration.swift
//
//  AlhPdfViewConfiguration.swift
//  alh_pdf_view
//
//  Created by André Börger on 29.03.22.
//

import Foundation

class AlhPdfViewConfiguration {
    var filePath: String?
    var bytes: Data?
    var enableSwipe: Bool
    var fitPolicy: FitPolicy
    var fitEachPage: Bool
    var swipeHorizontal: Bool
    var password: String
    var nightMode: Bool
    var autoSpacing: Bool
    var pageFling: Bool
    var pageSnap: Bool
    var enableDoubleTap: Bool
    var defaultPage: Int
    var defaultZoomFactor: CGFloat
    var backgroundColor: UIColor
    var minZoom: CGFloat
    var maxZoom: CGFloat
    
    init(arguments: Dictionary<String, Any>) {
        self.filePath = arguments["filePath"] as? String
        
        if let flutterBytes = arguments["bytes"] as? FlutterStandardTypedData {
            self.bytes = flutterBytes.data
        }
    
        self.enableSwipe = arguments["enableSwipe"] as! Bool
        self.fitPolicy = FitPolicy.init(value: arguments["fitPolicy"] as! String)
        self.fitEachPage = arguments["fitEachPage"] as! Bool
        self.swipeHorizontal = arguments["swipeHorizontal"] as! Bool
        self.password = arguments["password"] as! String
        self.nightMode = arguments["nightMode"] as! Bool
        self.autoSpacing = arguments["autoSpacing"] as! Bool
        self.pageFling = arguments["pageFling"] as! Bool
        self.pageSnap = arguments["pageSnap"] as! Bool
        self.enableDoubleTap = arguments["enableDoubleTap"] as! Bool
        self.defaultPage = arguments["defaultPage"] as! Int
        self.defaultZoomFactor = arguments["defaultZoomFactor"] as! CGFloat
        self.backgroundColor = (arguments["backgroundColor"] as! UInt).toUIColor()
        self.minZoom = arguments["minZoom"] as! CGFloat
        self.maxZoom = arguments["maxZoom"] as! CGFloat
    }
}

enum FitPolicy: String {
    case both, height, width
    
    init(value: String) {
        switch(value) {
            case "FitPolicy.both": self =  FitPolicy.both
            case "FitPolicy.width": self =  FitPolicy.width
            case "FitPolicy.height": self =  FitPolicy.height
            default: self = FitPolicy.both
        }
    }
}

File: android/src/main/kotlin/com/example/alh_pdf_view/model/AlhPdfViewConfiguration.kt
package com.example.alh_pdf_view.model

import com.github.barteksc.pdfviewer.util.FitPolicy

class AlhPdfViewConfiguration(
    val filePath: String?,
    val bytes: ByteArray?,
    val autoSpacing: Boolean,
    val enableSwipe: Boolean,
    val fitEachPage: Boolean,
    val swipeHorizontal: Boolean,
    val password: String,
    val nightMode: Boolean,
    val pageFling: Boolean,
    val pageSnap: Boolean,
    val defaultPage: Int,
    val defaultZoomFactor: Double,
    val fitPolicy: FitPolicy,
    val backgroundColor: Int,
    val enableDoubleTap: Boolean,
    val minZoom: Float,
    val maxZoom: Float,
    val enableDefaultScrollHandle: Boolean
) {
    companion object {
        fun fromArguments(map: Map<*, *>): AlhPdfViewConfiguration {
            val filePath: String? = map["filePath"] as String?
            val bytes: ByteArray? = map["bytes"] as ByteArray?
            val autoSpacing: Boolean = map["autoSpacing"] as Boolean
            val enableSwipe: Boolean = map["enableSwipe"] as Boolean
            val fitEachPage: Boolean = map["fitEachPage"] as Boolean
            val swipeHorizontal: Boolean = map["swipeHorizontal"] as Boolean
            val password: String = map["password"] as String
            val nightMode: Boolean = map["nightMode"] as Boolean
            val pageFling: Boolean = map["pageFling"] as Boolean
            val pageSnap: Boolean = map["pageSnap"] as Boolean
            val defaultPage: Int = map["defaultPage"] as Int
            val defaultZoomFactor: Double = map["defaultZoomFactor"] as Double
            val fitPolicy: FitPolicy = fitPolicyFrom(map["fitPolicy"] as String)
            val backgroundColor: Int =
                if (map["backgroundColor"] is Int) (map["backgroundColor"] as Int) else (map["backgroundColor"] as Long).toInt()
            val enableDoubleTap = map["enableDoubleTap"] as Boolean
            val minZoom = map["minZoom"] as Double
            val maxZoom = map["maxZoom"] as Double
            val enableDefaultScrollHandle = map["enableDefaultScrollHandle"] as Boolean

            return AlhPdfViewConfiguration(
                filePath,
                bytes,
                autoSpacing,
                enableSwipe,
                fitEachPage,
                swipeHorizontal,
                password,
                nightMode,
                pageFling,
                pageSnap,
                defaultPage,
                defaultZoomFactor,
                fitPolicy,
                backgroundColor,
                enableDoubleTap,
                minZoom.toFloat(),
                maxZoom.toFloat(),
                enableDefaultScrollHandle
            )
        }

        /**
         * Looking for the correct [FitPolicy] depending on the value.
         *
         * Returns the default value [FitPolicy.BOTH] if no matching [String] was found.
         */
        private fun fitPolicyFrom(value: String): FitPolicy {
            return when (value) {
                "FitPolicy.width" -> FitPolicy.WIDTH
                "FitPolicy.height" -> FitPolicy.HEIGHT
                else -> FitPolicy.BOTH
            }
        }
    }
}